onnxruntime/core/providers/cuda/math/matmul.cc

```diff
@@ -88,6 +95,366 @@ static bool CanUseStridedBatchedGemm(const TensorShape& left_shape, const Tensor
+    }
+  }
+
+  // `cublasltmatmul` computes the following formula: D = alpha*(A*B) + beta*(C).
+  // Our matrix multiplication doesn't use a beta * C bias, so we just set the beta to 0 when calling the API.
+  // https://docs.nvidia.com/cuda/cublas/index.html?highlight=cublasLtMatmul#cublasltmatmul
+  float beta = 0;
+
+  const void* p_input_a = right_X_fp8.get();
+  const void* p_input_b = left_X_fp8.get();
+  const void* p_input_c = C.get();
+  void* p_output_y = Y->MutableDataRaw();
+
+  // Create matrix descriptors. Not setting any extra attributes.
+  // ORT is row_major, but cublas is col_major, so we need to use right_X as A and left_X as B.
+  int32_t dtype_A = right_X->GetElementType();
+  int32_t dtype_B = left_X->GetElementType();
+  int32_t dtype_C = Y->GetElementType();
+  int32_t dtype_Y = Y->GetElementType();
+  cudaDataType_t a_cuda_type = onnxruntime::cuda::ToCudaDataType(dtype_A);
+  cudaDataType_t b_cuda_type = onnxruntime::cuda::ToCudaDataType(dtype_B);
+  cudaDataType_t c_cuda_type = onnxruntime::cuda::ToCudaDataType(dtype_C);
+  cudaDataType_t y_cuda_type = onnxruntime::cuda::ToCudaDataType(dtype_Y);
+
+  // https://docs.nvidia.com/cuda/cublas/index.html?highlight=cublasStatus_t#cublasltmatmul
+  // "The compute type must be CUBLAS_COMPUTE_32F."
+  // "The scale type must be CUDA_R_32F."
+  cudaDataType_t scale_cuda_type = CUDA_R_32F;
+  cublasComputeType_t compute_type = CUBLAS_COMPUTE_32F;
+
+  cublasLtMatmulDesc_t operationDesc = nullptr;
+  CUBLAS_RETURN_IF_ERROR(cublasLtMatmulDescCreate(&operationDesc, compute_type, scale_cuda_type));
+  CUBLAS_RETURN_IF_ERROR(cublasLtMatmulDescSetAttribute(operationDesc, CUBLASLT_MATMUL_DESC_TRANSA, &transA, sizeof(transA)));
+  CUBLAS_RETURN_IF_ERROR(cublasLtMatmulDescSetAttribute(operationDesc, CUBLASLT_MATMUL_DESC_TRANSB, &transB, sizeof(transB)));
+
+  cublasLtMatrixLayout_t Adesc = nullptr, Bdesc = nullptr, Cdesc = nullptr, Ydesc = nullptr;
+  CUBLAS_RETURN_IF_ERROR(cublasLtMatrixLayoutCreate(&Adesc, a_cuda_type, K, N, lda)); // right_X
+  CUBLAS_RETURN_IF_ERROR(cublasLtMatrixLayoutCreate(&Bdesc, b_cuda_type, K, M, ldb)); // left_X
+  CUBLAS_RETURN_IF_ERROR(cublasLtMatrixLayoutCreate(&Ydesc, y_cuda_type, N, M, ldc)); // cublas is col major
+
+  int64_t sm_count_ = device_prop.multiProcessorCount;
+#if CUDA_VERSION >= 11060
+  // CUBLASLT_MATMUL_DESC_SM_COUNT_TARGET exists from https://docs.nvidia.com/cuda/archive/11.6.0/pdf/CUBLAS_Library.pdf
+  if (sm_count_ != 0) {
+    int math_sm_count = static_cast<int>(sm_count_);
+    CUBLAS_RETURN_IF_ERROR(cublasLtMatmulDescSetAttribute(
+        operationDesc, CUBLASLT_MATMUL_DESC_SM_COUNT_TARGET, &math_sm_count,
+        sizeof(math_sm_count)));
+  }
+#endif
+
+  // matmul float 8
+#if CUDA_VERSION >= 11080
+  float* quant_float = (float*)malloc(256 * sizeof(float));
+  for (int i = 0; i < 256; i ++) {
+    quant_float[i] = i;
+  }
+  float std_quant = ComputeStandardDeviation(quant_float, 256);
+  free(quant_float);
```

Comment:
quant_float is const vector, which means std_quant can be a constant. Why do we need compute it online?